# NYC Crime Dataset Cleaning
The purpose of this notebook is to clean the NYC crime dataset prior to EDA and project work. This includes removing unecessary characters/quotations around cell values, replacing empty rows with null, removing rows with missing data, and renaming columns. 

#### Library Imports
The following are all the libraries used for the data cleaning process

In [ ]:
#library imports
import pandas as pd
import numpy as np

#### Data View
Viewing what the data looks like to understand what needs cleaning.

In [6]:
data_df = pd.read_csv('nyc_crime_data.csv')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 10)
data_df.head(10)

C:\Users\Nicol\AppData\Local\Temp\ipykernel_12392\1590182787.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  data_df = pd.read_csv('nyc_crime_data.csv')


,CMPLNT_NUM,CMPLNT_FR_DT,CMPLNT_FR_TM,CMPLNT_TO_DT,CMPLNT_TO_TM,ADDR_PCT_CD,RPT_DT,KY_CD,OFNS_DESC,PD_CD,PD_DESC,CRM_ATPT_CPTD_CD,LAW_CAT_CD,BORO_NM,LOC_OF_OCCUR_DESC,PREM_TYP_DESC,JURIS_DESC,JURISDICTION_CODE,PARKS_NM,HADEVELOPT,HOUSING_PSA,X_COORD_CD,Y_COORD_CD,SUSP_AGE_GROUP,SUSP_RACE,SUSP_SEX,TRANSIT_DISTRICT,Latitude,Longitude,Lat_Lon,PATROL_BORO,STATION_NAME,VIC_AGE_GROUP,VIC_RACE,VIC_SEX
0,169450128,09/11/2017,16:30:00,09/13/2017,16:45:00,45,09/17/2017,578,HARRASSMENT 2,638.0,"HARASSMENT,SUBD 3,4,5",COMPLETED,VIOLATION,BRONX,FRONT OF,STREET,N.Y. POLICE DEPT,0,(null),(null),(null),"1,031,237","257,081",25-44,BLACK,U,NaN,40.872179,-73.830101,"(40.872179, -73.830101)",PATROL BORO BRONX,(null),45-64,BLACK,M
1,167086123,07/15/2017,21:15:00,07/15/2017,21:25:00,34,07/15/2017,106,FELONY ASSAULT,109.0,"ASSAULT 2,1,UNCLASSIFIED",COMPLETED,FELONY,MANHATTAN,INSIDE,RESIDENCE - APT. HOUSE,N.Y. POLICE DEPT,0,(null),(null),(null),"1,002,264","250,413",25-44,WHITE,M,NaN,40.853985,-73.934883,"(40.853985, -73.934883)",PATROL BORO MAN NORTH,(null),25-44,WHITE,M
2,162097170,01/28/2017,17:36:00,NaN,(null),20,02/20/2017,121,CRIMINAL MISCHIEF & RELATED OF,273.0,"TAMPERING 1,CRIMINAL",COMPLETED,FELONY,MANHATTAN,(null),TRANSIT - NYC SUBWAY,N.Y. TRANSIT POLICE,1,(null),(null),(null),"992,081","224,190",(null),(null),(null),1.0,40.782022,-73.971723,"(40.78202248, -73.97172308)",PATROL BORO MAN NORTH,81 ST.-MUSEUM OF NATURAL HISTO,UNKNOWN,UNKNOWN,E
3,161472604,01/31/2017,10:00:00,NaN,(null),75,01/31/2017,578,HARRASSMENT 2,638.0,"HARASSMENT,SUBD 3,4,5",COMPLETED,VIOLATION,BROOKLYN,INSIDE,RESIDENCE - APT. HOUSE,N.Y. POLICE DEPT,0,(null),(null),(null),"1,018,468","184,404",(null),(null),(null),NaN,40.672758,-73.876644,"(40.672758, -73.876644)",PATROL BORO BKLYN NORTH,(null),65+,BLACK,M
4,171196566,11/03/2017,19:48:00,11/03/2017,19:50:00,17,11/03/2017,344,ASSAULT 3 & RELATED OFFENSES,101.0,ASSAULT 3,COMPLETED,MISDEMEANOR,MANHATTAN,FRONT OF,STREET,N.Y. POLICE DEPT,0,(null),(null),(null),"993,039","215,158",65+,BLACK,M,NaN,40.757232,-73.968273,"(40.757232, -73.968273)",PATROL BORO MAN SOUTH,(null),45-64,WHITE HISPANIC,M
5,178601669,03/25/2018,14:00:00,04/25/2018,17:00:00,70,04/25/2018,109,GRAND LARCENY,439.0,"LARCENY,GRAND FROM OPEN AREAS, UNATTENDED",COMPLETED,FELONY,BROOKLYN,INSIDE,MAILBOX OUTSIDE,N.Y. POLICE DEPT,0,(null),(null),(null),"994,040","173,452",UNKNOWN,UNKNOWN,U,NaN,40.642758,-73.964721,"(40.642758, -73.964721)",PATROL BORO BKLYN SOUTH,(null),65+,BLACK,F
6,175311181,12/26/2017,19:00:00,12/30/2017,15:00:00,52,02/28/2018,109,GRAND LARCENY,422.0,"LARCENY,GRAND BY OPEN CREDIT CARD (NEW ACCT)",COMPLETED,FELONY,BRONX,INSIDE,RESIDENCE - APT. HOUSE,N.Y. POLICE DEPT,0,(null),(null),(null),"1,018,464","257,578",UNKNOWN,UNKNOWN,U,NaN,40.873602,-73.876285,"(40.873602, -73.876285)",PATROL BORO BRONX,(null),18-24,WHITE HISPANIC,F
7,190022223,11/06/2018,21:00:00,11/06/2018,21:30:00,102,11/14/2018,112,THEFT-FRAUD,739.0,"FRAUD,UNCLASSIFIED-FELONY",COMPLETED,FELONY,QUEENS,INSIDE,RESIDENCE - APT. HOUSE,N.Y. POLICE DEPT,0,(null),(null),(null),"1,028,107","194,436",(null),(null),(null),NaN,40.700251,-73.841828,"(40.700251, -73.841828)",PATROL BORO QUEENS SOUTH,(null),25-44,WHITE HISPANIC,F
8,172920554,12/22/2017,22:51:00,NaN,(null),14,12/23/2017,121,CRIMINAL MISCHIEF & RELATED OF,269.0,"MISCHIEF,CRIMINAL, UNCL 2ND",COMPLETED,FELONY,MANHATTAN,INSIDE,GYM/FITNESS FACILITY,N.Y. POLICE DEPT,0,(null),(null),(null),"987,816","211,442",UNKNOWN,WHITE HISPANIC,M,NaN,40.747036,-73.987127,"(40.747036, -73.987127)",PATROL BORO MAN SOUTH,(null),45-64,AMERICAN INDIAN/ALASKAN NATIVE,M
9,164382072,05/03/2017,01:50:00,05/03/2017,02:00:00,34,05/03/2017,344,ASSAULT 3 & RELATED OFFENSES,101.0,ASSAULT 3,COMPLETED,MISDEMEANOR,MANHATTAN,INSIDE,RESIDENCE - APT. HOUSE,N.Y. POLICE DEPT,0,(null),(null),(null),"1,003,184","252,167",(null),(null),(null),NaN,40.858795,-73.931549,"(40.858795, -73.931549)",PATROL BORO MAN NORTH,(null),25-44,WHITE,F


#### Data Observations Part 1

* `(null)` appears as a literal string value for some columns (e.g. BORO_NM, PARKS_NM, CMPLNT_TO_TM, SUSP_AGE_GROUP, VIC_SEX, etc.) but these need to be changed to actual null values
* Some data/time columns are stored as a literal string value as well (e.g., CMPLNT_FR_DT, CMPLNT_TO_DT, RPT_DT, CMPLNT_FR_TM, CMPLNT_TO_TM) but should be stored as proper date, time, or datetime values
* Some of these date/time columns should be combined to get a proper date-time time stamp too (e.g., CMPLNT_FR_DT and CMPLNT_FR_TM)
* The coordinate columns (X_COORD_CD and Y_COORD_CD) contain values like "1,031,237" but these commas make them strings instead of numbers so the commas need to be stripped and the remaining string should be turned to numeric

#### Data Summary
Summary of columns and the data types, and different column values found for each

In [7]:
def summarize_df(df, max_unique=15):
    rows = []
    for col in df.columns:
        series = df[col]
        total = len(series)
        
        null_count = series.isna().sum()
        # also catch string "(null)" values
        if series.dtype == object or pd.api.types.is_string_dtype(series):
            null_count += (series == '(null)').sum()
        pct_null = round(null_count / total * 100, 1)
        
        # data types found in column (handles mixed types)
        type_counts = series.dropna().map(type).value_counts()
        types_found = ", ".join([f"{t.__name__}({c})" for t, c in type_counts.items()])
        
        # unique values
        n_unique = series.nunique(dropna=True)
        if n_unique <= max_unique:
            sample_vals = series.dropna().unique().tolist()
            # remove string nulls
            sample_vals = [v for v in sample_vals if v != '(null)']
            values_summary = str(sample_vals)
        else:
            values_summary = f"{n_unique} distinct values"
        
        rows.append({
            "Column":        col,
            "Dtype":         str(series.dtype),
            "Types in Data": types_found,
            "Null/Empty":    f"{null_count} ({pct_null}%)",
            "Unique Count":  n_unique,
            "Values":        values_summary,
        })
    
    summary = pd.DataFrame(rows)
    return summary.style.set_properties(**{"text-align": "left"}) \
                        .set_table_styles([{"selector": "th", "props": [("text-align", "left")]}])

summarize_df(data_df)

,Column,Dtype,Types in Data,Null/Empty,Unique Count,Values
0,CMPLNT_NUM,object,"str(49152), int(848)",0 (0.0%),50000,50000 distinct values
1,CMPLNT_FR_DT,object,str(50000),0 (0.0%),730,730 distinct values
2,CMPLNT_FR_TM,object,str(50000),0 (0.0%),1416,1416 distinct values
3,CMPLNT_TO_DT,object,str(42606),7394 (14.8%),935,935 distinct values
4,CMPLNT_TO_TM,object,str(50000),7372 (14.7%),1430,1430 distinct values
5,ADDR_PCT_CD,int64,int(50000),0 (0.0%),77,77 distinct values
6,RPT_DT,object,str(50000),0 (0.0%),1042,1042 distinct values
7,KY_CD,int64,int(50000),0 (0.0%),58,58 distinct values
8,OFNS_DESC,object,str(50000),1 (0.0%),54,54 distinct values
9,PD_CD,float64,float(49970),30 (0.1%),287,287 distinct values


#### Data Observations Part 2

* 